### DISCUSSION TITLE
rbatis::Error::Database 能否支持传递SQL_STATE

### DISCUSSION THREAD

**Discussion by bin-liu** [rbatis/rbatis#260] [General]
SQLx中的error是包含SQL_STATE的，但rbatis::Error在转换时丢掉了。能否保留？
部分场景需要判断如Duplicate Key这样的错误做特殊处理。
如果可以建议Database错误能细化下，当前都是Database同一个类型的错误。

SQLx:
```rust
pub enum Error {
    ...
    /// Error returned from the database.
    #[error("error returned from database: {0}")]
    Database(#[source] Box<dyn DatabaseError>),
    ...
}
/// An error that was returned from the database.
pub trait DatabaseError: 'static + Send + Sync + StdError {
    ...
    /// The (SQLSTATE) code for the error.
    fn code(&self) -> Option<Cow<'_, str>> {
        None
    }
    ...
}
```

Rbatis:

```rust
...
impl From<sqlx_core::error::BoxDynError> for crate::Error {
    fn from(arg: BoxDynError) -> Self {
        return crate::Error::Database(arg.to_string());
    }
}

impl From<sqlx_core::error::Error> for crate::Error {
    fn from(arg: sqlx_core::error::Error) -> Self {
        return crate::Error::Database(arg.to_string());
    }
}
...
```

**Comment by zhuxiujia** [comment]
> SQLx中的error是包含SQL_STATE的，但rbatis::Error在转换时丢掉了。能否保留？ 部分场景需要判断如Duplicate Key这样的错误做特殊处理。 如果可以建议Database错误能细化下，当前都是Database同一个类型的错误。
> 
> SQLx:
> 
> ```rust
> pub enum Error {
>     ...
>     /// Error returned from the database.
>     #[error("error returned from database: {0}")]
>     Database(#[source] Box<dyn DatabaseError>),
>     ...
> }
> /// An error that was returned from the database.
> pub trait DatabaseError: 'static + Send + Sync + StdError {
>     ...
>     /// The (SQLSTATE) code for the error.
>     fn code(&self) -> Option<Cow<'_, str>> {
>         None
>     }
>     ...
> }
> ```
> 
> Rbatis:
> 
> ```rust
> ...
> impl From<sqlx_core::error::BoxDynError> for crate::Error {
>     fn from(arg: BoxDynError) -> Self {
>         return crate::Error::Database(arg.to_string());
>     }
> }
> 
> impl From<sqlx_core::error::Error> for crate::Error {
>     fn from(arg: sqlx_core::error::Error) -> Self {
>         return crate::Error::Database(arg.to_string());
>     }
> }
> ...
> ```

to_string应该携带信息的，应该是sqlx丢失了没有进入to_string里面的文本内容
